cpp/include/raft/sparse/matrix/detail/preprocessing.cuh

```diff
@@ -0,0 +1,215 @@
+                                              raft::device_vector_view<T2, int64_t> doc_lengths,
+                                              raft::device_vector_view<T2, int64_t> term_counts)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+
+  auto num_rows =
+    raft::sparse::neighbors::get_n_components(rows.data_handle(), rows.size(), stream);
+
+  auto row_keys   = raft::make_device_vector<int, int64_t>(handle, num_rows);
+  auto row_counts = raft::make_device_vector<float, int64_t>(handle, num_rows);
+  auto row_fill   = raft::make_device_vector<float, int64_t>(handle, rows.size());
+
+  // the amount of columns(documents) that each row(term) is found in
+  thrust::fill(raft::resource::get_thrust_policy(handle),
+               row_fill.data_handle(),
+               row_fill.data_handle() + row_fill.size(),
+               1.0f);
+  get_uniques_counts(
+    handle, rows, columns, values, row_fill.view(), row_keys.view(), row_counts.view());
+
+  create_mapped_vector<int, float>(handle, rows, row_keys.view(), row_counts.view(), term_counts);
+  auto num_cols =
+    raft::sparse::neighbors::get_n_components(columns.data_handle(), columns.size(), stream);
+  auto col_keys   = raft::make_device_vector<int, int64_t>(handle, num_cols);
+  auto col_counts = raft::make_device_vector<float, int64_t>(handle, num_cols);
+
+  get_uniques_counts(handle, columns, rows, values, values, col_keys.view(), col_counts.view());
+
+  int total_document_lengths = thrust::reduce(raft::resource::get_thrust_policy(handle),
+                                              col_counts.data_handle(),
+                                              col_counts.data_handle() + col_counts.size());
+  float avg_doc_length       = float(total_document_lengths) / col_keys.size();
+
+  create_mapped_vector<int, float>(
+    handle, columns, col_keys.view(), col_counts.view(), doc_lengths);
+  return {col_keys.size(), avg_doc_length};
+}
+
+template <typename T1, typename T2>
+void encode_tfidf(raft::resources& handle,
+                  raft::device_vector_view<T1, int64_t> rows,
+                  raft::device_vector_view<T1, int64_t> columns,
+                  raft::device_vector_view<T2, int64_t> values,
+                  raft::device_vector_view<T2, int64_t> values_out)
+{
+  auto doc_lengths = raft::make_device_vector<float, int64_t>(handle, columns.size());
+  auto term_counts = raft::make_device_vector<float, int64_t>(handle, rows.size());
+  auto [doc_count, avg_doc_length] = sparse_search_preprocess<int, float>(
+    handle, rows, columns, values, doc_lengths.view(), term_counts.view());
+
+  raft::linalg::map(handle,
+                    values_out,
+                    tfidf(doc_count),
+                    raft::make_const_mdspan(values),
+                    raft::make_const_mdspan(term_counts.view()));
+}
+
+template <typename T1, typename T2>
+void encode_bm25(raft::resources& handle,
```

Comment:
might be better to create a single encode and pass the desired struct via a function parameter that can be relayed to the map call.